File: swim-system-java/swim-core-java/swim.collections/src/main/java/swim/collections/STreeNode.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.collections;

import java.util.Map;
import swim.util.Cursor;

final class STreeNode<T> extends STreePage<T> {
  final STreePage<T>[] pages;
  final int[] knots;
  final int size;

  STreeNode(STreePage<T>[] pages, int[] knots, int size) {
    this.pages = pages;
    if (knots == null || size < 0) {
      knots = new int[pages.length - 1];
      size = 0;
      for (int i = 0, n = knots.length; i < n; i += 1) {
        size += pages[i].size();
        knots[i] = size;
      }
      size += pages[knots.length].size();
    }
    this.knots = knots;
    this.size = size;
  }

  STreeNode(STreePage<T>[] pages) {
    this(pages, null, -1);
  }

  @Override
  public boolean isEmpty() {
    return this.size == 0;
  }

  @Override
  public int size() {
    return this.size;
  }

  @Override
  public int arity() {
    return this.pages.length;
  }

  @Override
  public boolean contains(Object value) {
    final STreePage<T>[] pages = this.pages;
    for (int i = 0, n = pages.length; i < n; i += 1) {
      if (pages[i].contains(value)) {
        return true;
      }
    }
    return false;
  }

  @Override
  public int indexOf(Object object) {
    final STreePage<T>[] pages = this.pages;
    int k = 0;
    for (int x = 0, n = pages.length; x < n; x += 1) {
      final STreePage<T> page = pages[x];
      final int i = page.indexOf(object);
      if (i >= 0) {
        return k + i;
      }
      k += page.size();
    }
    return -1;
  }

  @Override
  public int lastIndexOf(Object object) {
    final STreePage<T>[] pages = this.pages;
    int k = this.size;
    for (int x = pages.length - 1; x >= 0; x -= 1) {
      final STreePage<T> page = pages[x];
      final int i = page.lastIndexOf(object);
      k -= page.size();
      if (i >= 0) {
        return k + 1;
      }
    }
    return -1;
  }

  @Override
  public T get(int index) {
    int x = lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    final int i = x == 0 ? index : index - this.knots[x - 1];
    return this.pages[x].get(i);
  }

  @Override
  public Map.Entry<Object, T> getEntry(int index) {
    int x = lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    final int i = x == 0 ? index : index - this.knots[x - 1];
    return this.pages[x].getEntry(i);
  }

  @Override
  public STreeNode<T> updated(int index, T newValue, STreeContext<T> tree) {
    int x = lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    final int i = x == 0 ? index : index - this.knots[x - 1];
    final STreePage<T> oldPage = this.pages[x];
    final STreePage<T> newPage = oldPage.updated(i, newValue, tree);
    if (oldPage != newPage) {
      if (oldPage.size() != newPage.size() && tree.pageShouldSplit(newPage)) {
        return updatedPageSplit(x, newPage, oldPage);
      } else {
        return updatedPage(x, newPage, oldPage);
      }
    } else {
      return this;
    }
  }

  @SuppressWarnings("unchecked")
  private STreeNode<T> updatedPage(int x, STreePage<T> newPage, STreePage<T> oldPage) {
    final STreePage<T>[] oldPages = this.pages;
    final int n = oldPages.length;
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
    System.arraycopy(oldPages, 0, newPages, 0, n);
    newPages[x] = newPage;

    final int[] oldKnots = this.knots;
    final int[] newKnots;
    int newSize;
    if (n - 1 > 0) {
      newKnots = new int[n - 1];
      if (x > 0) {
        System.arraycopy(oldKnots, 0, newKnots, 0, x);
        newSize = oldKnots[x - 1];
      } else {
        newSize = 0;
      }
      for (int i = x; i < n - 1; i += 1) {
        newSize += newPages[i].size();
        newKnots[i] = newSize;
      }
      newSize += newPages[n - 1].size();
    } else {
      newKnots = new int[0];
      newSize = 0;
    }

    return new STreeNode<T>(newPages, newKnots, newSize);
  }

  @SuppressWarnings("unchecked")
  private STreeNode<T> updatedPageSplit(int x, STreePage<T> newPage, STreePage<T> oldPage) {
    final STreePage<T>[] oldPages = this.pages;
    final int n = oldPages.length + 1;
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
    System.arraycopy(oldPages, 0, newPages, 0, x);

    final int y = newPage.arity() >>> 1;
    final STreePage<T> newLeftPage = newPage.splitLeft(y);
    final STreePage<T> newRightPage = newPage.splitRight(y);
    newPages[x] = newLeftPage;
    newPages[x + 1] = newRightPage;
    System.arraycopy(oldPages, x + 1, newPages, x + 2, n - (x + 2));

    return new STreeNode<T>(newPages);
  }

  @SuppressWarnings("unchecked")
  private STreeNode<T> updatedPageMerge(int x, STreeNode<T> newPage, STreePage<T> oldPage) {
    final STreePage<T>[] oldPages = this.pages;
    final STreePage<T>[] midPages = newPage.pages;
    final int k = midPages.length;
    final int n = oldPages.length + (k - 1);
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
    System.arraycopy(oldPages, 0, newPages, 0, x);
    System.arraycopy(midPages, 0, newPages, x, k);
    System.arraycopy(oldPages, x + 1, newPages, x + k, n - (x + k));

    return new STreeNode<T>(newPages);
  }

  @Override
  public STreeNode<T> inserted(int index, T newValue, Object id, STreeContext<T> tree) {
    int x = lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    final int i = x == 0 ? index : index - this.knots[x - 1];
    final STreePage<T> oldPage = this.pages[x];
    final STreePage<T> newPage = oldPage.inserted(i, newValue, id, tree);
    if (oldPage != newPage) {
      if (tree.pageShouldSplit(newPage)) {
        return updatedPageSplit(x, newPage, oldPage);
      } else {
        return updatedPage(x, newPage, oldPage);
      }
    } else {
      return this;
    }
  }

  @Override
  public STreePage<T> removed(int index, STreeContext<T> tree) {
    int x = lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    final int i = x == 0 ? index : index - this.knots[x - 1];
    final STreePage<T> oldPage = this.pages[x];
    final STreePage<T> newPage = oldPage.removed(i, tree);
    if (oldPage != newPage) {
      return replacedPage(x, newPage, oldPage, tree);
    } else {
      return this;
    }
  }

  private STreePage<T> replacedPage(int x, STreePage<T> newPage,
                                    STreePage<T> oldPage, STreeContext<T> tree) {
    if (!newPage.isEmpty()) {
      if (newPage instanceof STreeNode<?> && tree.pageShouldMerge(newPage)) {
        return updatedPageMerge(x, (STreeNode<T>) newPage, oldPage);
      } else {
        return updatedPage(x, newPage, oldPage);
      }
    } else if (this.pages.length > 2) {
      return removedPage(x, newPage, oldPage);
    } else if (this.pages.length > 1) {
      if (x == 0) {
        return this.pages[1];
      } else {
        return this.pages[0];
      }
    } else {
      return STreeLeaf.empty();
    }
  }

  @SuppressWarnings("unchecked")
  private STreeNode<T> removedPage(int x, STreePage<T> newPage, STreePage<T> oldPage) {
    final STreePage<T>[] oldPages = this.pages;
    final int n = oldPages.length - 1;
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
    System.arraycopy(oldPages, 0, newPages, 0, x);
    System.arraycopy(oldPages, x + 1, newPages, x, n - x);

    final int[] oldKnots = this.knots;
    final int[] newKnots = new int[n - 1];
    int newSize;
    if (x > 0) {
      System.arraycopy(oldKnots, 0, newKnots, 0, x);
      newSize = oldKnots[x - 1];
    } else {
      newSize = 0;
    }
    for (int i = x; i < n - 1; i += 1) {
      newSize += newPages[i].size();
      newKnots[i] = newSize;
    }
    newSize += newPages[n - 1].size();

    return new STreeNode<T>(newPages, newKnots, newSize);
  }

  @Override
  public STreePage<T> removed(Object value, STreeContext<T> tree) {
    final STreePage<T>[] pages = this.pages;
    for (int x = 0, n = pages.length; x < n; x += 1) {
      final STreePage<T> oldPage = pages[x];
      final STreePage<T> newPage = oldPage.removed(value, tree);
      if (oldPage != newPage) {
        return replacedPage(x, newPage, oldPage, tree);
      }
    }
    return this;
  }

  @SuppressWarnings("unchecked")
  @Override
  public STreePage<T> drop(int lower, STreeContext<T> tree) {
    if (lower > 0) {
      if (lower < this.size) {
        int x = lookup(lower);
        if (x >= 0) {
          x += 1;
        } else {
          x = -(x + 1);
        }
        final int i = x == 0 ? lower : lower - this.knots[x - 1];
        final STreePage<T>[] oldPages = this.pages;
        final int k = oldPages.length;
        final int n = k - x;
        if (n > 1) {
          final STreeNode<T> newNode;
          if (x > 0) {
            final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
            System.arraycopy(oldPages, x, newPages, 0, n);
            newNode = new STreeNode<T>(newPages);
          } else {
            newNode = this;
          }
          if (i > 0) {
            final STreePage<T> oldPage = oldPages[x];
            final STreePage<T> newPage = oldPage.drop(i, tree);
            return newNode.replacedPage(0, newPage, oldPage, tree);
          } else {
            return newNode;
          }
        } else {
          return oldPages[x].drop(i, tree);
        }
      } else {
        return STreeLeaf.empty();
      }
    } else {
      return this;
    }
  }

  @SuppressWarnings("unchecked")
  @Override
  public STreePage<T> take(int upper, STreeContext<T> tree) {
    if (upper < this.size) {
      if (upper > 0) {
        int x = lookup(upper);
        if (x >= 0) {
          x += 1;
        } else {
          x = -(x + 1);
        }
        final int i = x == 0 ? upper : upper - this.knots[x - 1];
        final STreePage<T>[] oldPages = this.pages;
        final int k = oldPages.length;
        final int n = i == 0 ? x : x + 1;
        if (n > 1) {
          final STreeNode<T> newNode;
          if (x < k) {
            final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[n];
            System.arraycopy(oldPages, 0, newPages, 0, n);
            final int[] newKnots = new int[n - 1];
            System.arraycopy(this.knots, 0, newKnots, 0, n - 1);
            final int newSize = newKnots[n - 2] + newPages[n - 1].size();
            newNode = new STreeNode<T>(newPages, newKnots, newSize);
          } else {
            newNode = this;
          }
          if (i > 0) {
            final STreePage<T> oldPage = oldPages[x];
            final STreePage<T> newPage = oldPage.take(i, tree);
            return newNode.replacedPage(x, newPage, oldPage, tree);
          } else {
            return newNode;
          }
        } else if (i > 0) {
          return oldPages[0].take(i, tree);
        } else {
          return oldPages[0];
        }
      } else {
        return STreeLeaf.empty();
      }
    } else {
      return this;
    }
  }

  @Override
  public STreeNode<T> balanced(STreeContext<T> tree) {
    if (this.pages.length > 1 && tree.pageShouldSplit(this)) {
      final int x = this.knots.length >>> 1;
      return split(x);
    } else {
      return this;
    }
  }

  @SuppressWarnings("unchecked")
  @Override
  public STreeNode<T> split(int x) {
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[2];
    final STreeNode<T> newLeftPage = splitLeft(x);
    final STreeNode<T> newRightPage = splitRight(x);
    newPages[0] = newLeftPage;
    newPages[1] = newRightPage;

    final int[] newKnots = new int[1];
    newKnots[0] = newLeftPage.size();

    return new STreeNode<T>(newPages, newKnots, this.size);
  }

  @SuppressWarnings("unchecked")
  @Override
  public STreeNode<T> splitLeft(int x) {
    final STreePage<T>[] oldPages = this.pages;
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[x + 1];
    System.arraycopy(oldPages, 0, newPages, 0, x + 1);

    final int[] oldKnots = this.knots;
    final int[] newKnots = new int[x];
    System.arraycopy(oldKnots, 0, newKnots, 0, x);

    int newSize = 0;
    for (int i = 0; i <= x; i += 1) {
      newSize += newPages[i].size();
    }

    return new STreeNode<T>(newPages, newKnots, newSize);
  }

  @SuppressWarnings("unchecked")
  @Override
  public STreeNode<T> splitRight(int x) {
    final STreePage<T>[] oldPages = this.pages;
    final int y = oldPages.length - (x + 1);
    final STreePage<T>[] newPages = (STreePage<T>[]) new STreePage<?>[y];
    System.arraycopy(oldPages, x + 1, newPages, 0, y);

    final int[] newKnots = new int[y - 1];
    int newSize;
    if (y > 0) {
      newSize = newPages[0].size();
      for (int i = 1; i < y; i += 1) {
        newKnots[i - 1] = newSize;
        newSize += newPages[i].size();
      }
    } else {
      newSize = 0;
    }

    return new STreeNode<T>(newPages, newKnots, newSize);
  }

  @Override
  public void copyToArray(Object[] array, int offset) {
    final STreePage<T>[] pages = this.pages;
    for (int x = 0, n = pages.length; x < n; x += 1) {
      final STreePage<T> page = pages[x];
      page.copyToArray(array, offset);
      offset += page.size();
    }
  }

  @Override
  public Cursor<Map.Entry<Object, T>> entryIterator() {
    return new STreeNodeCursor<T>(this.pages);
  }

  @Override
  public Cursor<Map.Entry<Object, T>> reverseEntryIterator() {
    return new STreeNodeCursor<T>(this.pages, this.size, this.pages.length);
  }

  private int lookup(int index) {
    int lo = 0;
    int hi = this.knots.length - 1;
    while (lo <= hi) {
      final int mid = (lo + hi) >>> 1;
      if (index > this.knots[mid]) {
        lo = mid + 1;
      } else if (index < this.knots[mid]) {
        hi = mid - 1;
      } else {
        return mid;
      }
    }
    return -(lo + 1);
  }
}

File: swim-js/swim-core/swim-collections/main/STreeNode.ts
// Copyright 2015-2023 Nstream, inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Cursor} from "@swim/util";
import {NodeCursor} from "./NodeCursor";
import type {STreeContext} from "./STreeContext";
import type {STree} from "./STree";
import {STreePage} from "./STreePage";

/** @internal */
export class STreeNode<V, I> extends STreePage<V, I> {
  constructor(pages: readonly STreePage<V, I>[], knots: readonly number[], size: number) {
    super();
    this.pages = pages;
    this.knots = knots;
    this.size = size;
  }

  /** @internal */
  readonly pages: readonly STreePage<V, I>[];

  /** @internal */
  readonly knots: readonly number[];

  override get arity(): number {
    return this.pages.length;
  }

  override readonly size: number;

  override isEmpty(): boolean {
    return this.size === 0;
  }

  override get(index: number): V | undefined {
    let x = this.lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    const i = x === 0 ? index : index - this.knots[x - 1]!;
    return this.pages[x]!.get(i);
  }

  override getEntry(index: number): [I, V] | undefined {
    let x = this.lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    const i = x === 0 ? index : index - this.knots[x - 1]!;
    return this.pages[x]!.getEntry(i);
  }

  override updated(index: number, newValue: V, tree: STreeContext<V, I>): STreeNode<V, I> {
    let x = this.lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    const i = x === 0 ? index : index - this.knots[x - 1]!;
    const oldPage = this.pages[x]!;
    const newPage = oldPage.updated(i, newValue, tree);
    if (oldPage === newPage) {
      return this;
    } else if (oldPage.size !== newPage.size && tree.pageShouldSplit(newPage)) {
      return this.updatedPageSplit(x, newPage, oldPage);
    }
    return this.updatedPage(x, newPage, oldPage);
  }

  /** @internal */
  updatedPage(x: number, newPage: STreePage<V, I>, oldPage: STreePage<V, I>): STreeNode<V, I> {
    const oldPages = this.pages;
    const newPages = oldPages.slice(0);
    newPages[x] = newPage;

    const oldKnots = this.knots;
    let newKnots: number[];
    let newSize: number;
    if (oldPages.length - 1 <= 0) {
      newKnots = [];
      newSize = 0;
    } else {
      newKnots = oldKnots.slice(0);
      if (x > 0) {
        newSize = oldKnots[x - 1]!;
      } else {
        newSize = 0;
      }
      for (let i = x; i < newKnots.length; i += 1) {
        newSize += newPages[i]!.size;
        newKnots[i] = newSize;
      }
      newSize += newPages[newKnots.length]!.size;
    }

    return new STreeNode(newPages, newKnots, newSize);
  }

  /** @internal */
  updatedPageSplit(x: number, newPage: STreePage<V, I>, oldPage: STreePage<V, I>): STreeNode<V, I> {
    const oldPages = this.pages;
    const newPages = new Array<STreePage<V, I>>(oldPages.length + 1);
    for (let i = 0; i < x; i += 1) {
      newPages[i] = oldPages[i]!;
    }

    const newLeftPage = newPage.splitLeft(newPage.arity >>> 1);
    const newRightPage = newPage.splitRight(newPage.arity >>> 1);
    newPages[x] = newLeftPage;
    newPages[x + 1] = newRightPage;
    for (let i = x + 1; i < oldPages.length; i += 1) {
      newPages[i + 1] = oldPages[i]!;
    }

    return STreeNode.create(newPages);
  }

  /** @internal */
  updatedPageMerge(x: number, newPage: STreeNode<V, I>, oldPage: STreePage<V, I>): STreeNode<V, I> {
    const oldPages = this.pages;
    const midPages = newPage.pages;
    const newPages = new Array<STreePage<V, I>>(oldPages.length + midPages.length - 1);
    for (let i = 0; i < x; i += 1) {
      newPages[i] = oldPages[i]!;
    }
    for (let i = 0; i < midPages.length; i += 1) {
      newPages[i + x] = midPages[i]!;
    }
    for (let i = x + 1; i < oldPages.length; i += 1) {
      newPages[i + midPages.length - 1] = oldPages[i]!;
    }

    return STreeNode.create(newPages);
  }

  override inserted(index: number, newValue: V, id: I | undefined, tree: STreeContext<V, I>): STreeNode<V, I> {
    let x = this.lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    const subIndex = x === 0 ? index : index - this.knots[x - 1]!;
    const oldPage = this.pages[x]!;
    const newPage = oldPage.inserted(subIndex, newValue, id, tree);
    if (oldPage === newPage) {
      return this;
    } else if (tree.pageShouldSplit(newPage)) {
      return this.updatedPageSplit(x, newPage, oldPage);
    }
    return this.updatedPage(x, newPage, oldPage);
  }

  override removed(index: number, tree: STreeContext<V, I>): STreePage<V, I> {
    let x = this.lookup(index);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    const subIndex = x === 0 ? index : index - this.knots[x - 1]!;
    const oldPage = this.pages[x]!;
    const newPage = oldPage.removed(subIndex, tree);
    if (oldPage === newPage) {
      return this;
    }
    return this.replacedPage(x, newPage, oldPage, tree);
  }

  /** @internal */
  replacedPage(x: number, newPage: STreePage<V, I>, oldPage: STreePage<V, I>,
               tree: STreeContext<V, I>): STreePage<V, I> {
    if (!newPage.isEmpty()) {
      if (newPage instanceof STreeNode && tree.pageShouldMerge(newPage)) {
        return this.updatedPageMerge(x, newPage, oldPage);
      }
      return this.updatedPage(x, newPage, oldPage);
    } else if (this.pages.length > 2) {
      return this.removedPage(x, newPage, oldPage);
    } else if (this.pages.length > 1) {
      if (x === 0) {
        return this.pages[1]!;
      } else {
        return this.pages[0]!;
      }
    }
    return STreePage.empty();
  }

  /** @internal */
  removedPage(x: number, newPage: STreePage<V, I>, oldPage: STreePage<V, I>): STreeNode<V, I> {
    const oldPages = this.pages;
    const newPages = new Array<STreePage<V, I>>(oldPages.length - 1);
    for (let i = 0; i < x; i += 1) {
      newPages[i] = oldPages[i]!;
    }
    for (let i = x + 1; i < oldPages.length; i += 1) {
      newPages[i - 1] = oldPages[i]!;
    }

    const oldKnots = this.knots;
    const newKnots = new Array<number>(oldKnots.length - 1);
    let newSize: number;
    if (x <= 0) {
      newSize = 0;
    } else {
      for (let i = 0; i < x; i += 1) {
        newKnots[i] = oldKnots[i]!;
      }
      newSize = oldKnots[x - 1]!;
    }
    for (let i = x; i < newKnots.length; i += 1) {
      newSize += newPages[i]!.size;
      newKnots[i] = newSize;
    }
    newSize += newPages[newKnots.length]!.size;

    return new STreeNode(newPages, newKnots, newSize);
  }

  override drop(lower: number, tree: STreeContext<V, I>): STreePage<V, I> {
    if (lower <= 0) {
      return this;
    } else if (lower >= this.size) {
      return STreePage.empty();
    }
    let x = this.lookup(lower);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    lower = x === 0 ? lower : lower - this.knots[x - 1]!;
    const oldPages = this.pages;
    const n = oldPages.length - x;
    if (n <= 1) {
      return oldPages[x]!.drop(lower, tree);
    }
    let newNode: STreeNode<V, I>;
    if (x <= 0) {
      newNode = this;
    } else {
      const newPages = new Array<STreePage<V, I>>(n);
      for (let i = 0; i < n; i += 1) {
        newPages[i] = oldPages[i + x]!;
      }
      newNode = STreeNode.create(newPages);
    }
    if (lower <= 0) {
      return newNode;
    }
    const oldPage = oldPages[x]!;
    const newPage = oldPage.drop(lower, tree);
    return newNode.replacedPage(0, newPage, oldPage, tree);
  }

  override take(upper: number, tree: STreeContext<V, I>): STreePage<V, I> {
    if (upper >= this.size) {
      return this;
    } else if (upper <= 0) {
      return STreePage.empty();
    }
    let x = this.lookup(upper);
    if (x >= 0) {
      x += 1;
    } else {
      x = -(x + 1);
    }
    upper = x === 0 ? upper : upper - this.knots[x - 1]!;
    const oldPages = this.pages;
    const n = upper === 0 ? x : x + 1;
    if (n <= 1) {
      if (upper > 0) {
        return oldPages[0]!.take(upper, tree);
      } else {
        return oldPages[0]!;
      }
    }
    let newNode: STreeNode<V, I>;
    if (x >= oldPages.length) {
      newNode = this;
    } else {
      const newPages = new Array<STreePage<V, I>>(n);
      for (let i = 0; i < n; i += 1) {
        newPages[i] = oldPages[i]!;
      }
      const newKnots = new Array<number>(n - 1);
      for (let i = 0; i < newKnots.length; i += 1) {
        newKnots[i] = this.knots[i]!;
      }
      const newSize = newKnots[n - 2]! + newPages[n - 1]!.size;
      newNode = new STreeNode(newPages, newKnots, newSize);
    }
    if (upper <= 0) {
      return newNode;
    }
    const oldPage = oldPages[x]!;
    const newPage = oldPage.take(upper, tree);
    return newNode.replacedPage(x, newPage, oldPage, tree);
  }

  override balanced(tree: STreeContext<V, I>): STreeNode<V, I> {
    if (this.pages.length <= 1 || !tree.pageShouldSplit(this)) {
      return this;
    }
    return this.split(this.knots.length >>> 1);
  }

  override split(x: number): STreeNode<V, I> {
    const newPages = new Array<STreePage<V, I>>(2);
    const newLeftPage = this.splitLeft(x);
    const newRightPage = this.splitRight(x);
    newPages[0] = newLeftPage;
    newPages[1] = newRightPage;

    const newKnots = new Array<number>(1);
    newKnots[0] = newLeftPage.size;

    return new STreeNode(newPages, newKnots, this.size);
  }

  override splitLeft(x: number): STreeNode<V, I> {
    const oldPages = this.pages;
    const newPages = new Array<STreePage<V, I>>(x + 1);
    for (let i = 0; i < x + 1; i += 1) {
      newPages[i] = oldPages[i]!;
    }

    const oldKnots = this.knots;
    const newKnots = new Array<number>(x);
    for (let i = 0; i < x; i += 1) {
      newKnots[i] = oldKnots[i]!;
    }

    let newSize = 0;
    for (let i = 0; i <= x; i += 1) {
      newSize += newPages[i]!.size;
    }

    return new STreeNode(newPages, newKnots, newSize);
  }

  override splitRight(x: number): STreeNode<V, I> {
    const oldPages = this.pages;
    const y = oldPages.length - (x + 1);
    const newPages = new Array<STreePage<V, I>>(y);
    for (let i = 0; i < y; i += 1) {
      newPages[i] = oldPages[i + (x + 1)]!;
    }

    const newKnots = new Array<number>(y - 1);
    let newSize;
    if (y <= 0) {
      newSize = 0;
    } else {
      newSize = newPages[0]!.size;
      for (let i = 1; i < y; i += 1) {
        newKnots[i - 1] = newSize;
        newSize += newPages[i]!.size;
      }
    }

    return new STreeNode(newPages, newKnots, newSize);
  }

  override forEach<T, S>(callback: (this: S, value: V, index: number, id: I, tree: STree<V, I>) => T | void,
                         thisArg: S, offset: number, tree: STree<V, I>): T | undefined {
    for (let i = 0; i < this.pages.length; i += 1) {
      const page = this.pages[i]!;
      const result = page.forEach(callback, thisArg, offset, tree);
      if (result !== void 0) {
        return result;
      }
      offset += page.size;
    }
    return void 0;
  }

  override entries(): Cursor<[I, V]> {
    return new STreeNodeCursor(this.pages);
  }

  override reverseEntries(): Cursor<[I, V]> {
    return new STreeNodeCursor(this.pages, this.size, this.pages.length);
  }

  /** @internal */
  lookup(index: number): number {
    let lo = 0;
    let hi = this.knots.length - 1;
    while (lo <= hi) {
      const mid = (lo + hi) >>> 1;
      if (index > this.knots[mid]!) {
        lo = mid + 1;
      } else if (index < this.knots[mid]!) {
        hi = mid - 1;
      } else {
        return mid;
      }
    }
    return -(lo + 1);
  }

  static create<V, I>(pages: readonly STreePage<V, I>[]): STreeNode<V, I> {
    const knots = new Array<number>(pages.length - 1);
    let size = 0;
    for (let i = 0, n  = knots.length; i < n; i += 1) {
      size += pages[i]!.size;
      knots[i] = size;
    }
    size += pages[knots.length]!.size;
    return new STreeNode(pages, knots, size);
  }
}

/** @internal */
export class STreeNodeCursor<V, I> extends NodeCursor<[I, V], STreePage<V, I>> {
  constructor(pages: readonly STreePage<V, I>[], index: number = 0,
              childIndex: number = 0, childCursor: Cursor<[I, V]> | null = null) {
    super(pages, index, childIndex, childCursor);
  }

  protected override pageSize(page: STreePage<V, I>): number {
    return page.size;
  }

  protected override pageCursor(page: STreePage<V, I>): Cursor<[I, V]> {
    return page.entries();
  }

  protected override reversePageCursor(page: STreePage<V, I>): Cursor<[I, V]> {
    return page.reverseEntries();
  }
}
